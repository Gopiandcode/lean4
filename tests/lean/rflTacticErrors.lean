
/-!
This file tests the `rfl` tactic:
 * Extensibility
 * Error messages
 * Effect of `with_reducible`
-/

-- Setup

inductive P : α → α → Prop where | refl : P a a
attribute [refl] P.refl

-- Defeq to relation `P` at reducible transparency
abbrev Q : α → α → Prop := P
-- Defeq to relation `P` at default transparency
def Q' : α → α → Prop := P

-- No refl attribute
inductive R : α → α → Prop where | refl : R a a

-- Syntactic equal

example : true = true   := by rfl
example : HEq true true := by rfl
example : True ↔ True   := by rfl
example : P true true   := by rfl
example : Q true true   := by rfl
example : Q' true true  := by rfl -- Error
example : R true true   := by rfl -- Error

example : true = true   := by with_reducible rfl
example : HEq true true := by with_reducible rfl
example : True ↔ True   := by with_reducible rfl
example : P true true   := by with_reducible rfl
example : Q true true   := by with_reducible rfl
example : Q' true true  := by with_reducible rfl -- Error
example : R true true   := by with_reducible rfl -- Error

-- Reducibly equal

abbrev true' := true
abbrev True' := True

example : true' = true   := by rfl
example : HEq true' true := by rfl
example : True' ↔ True   := by rfl
example : P true' true   := by rfl
example : Q true' true   := by rfl
example : Q' true' true  := by rfl -- Error
example : R true' true   := by rfl -- Error

example : true' = true   := by with_reducible rfl
example : HEq true' true := by with_reducible rfl
example : True' ↔ True   := by with_reducible rfl
example : P true' true   := by with_reducible rfl
example : Q true' true   := by with_reducible rfl -- NB: No error!
example : Q' true' true  := by with_reducible rfl -- Error
example : R true' true   := by with_reducible rfl -- Error

-- Equal at default transparency only

def true'' := true
def True'' := True

example : true'' = true   := by rfl
example : HEq true'' true := by rfl
example : True'' ↔ True   := by rfl
example : P true'' true   := by rfl
example : Q true'' true   := by rfl
example : Q' true'' true  := by rfl -- Error
example : R true'' true   := by rfl -- Error

example : true'' = true   := by with_reducible rfl -- Error
example : HEq true'' true := by with_reducible rfl -- Error
example : True'' ↔ True   := by with_reducible rfl -- Error
example : P true'' true   := by with_reducible rfl -- Error
example : Q true'' true   := by with_reducible rfl -- Error
example : Q' true'' true  := by with_reducible rfl -- Error
example : R true'' true   := by with_reducible rfl -- Error

-- Unequal
example : false = true   := by rfl -- Error
example : HEq false true := by rfl -- Error
example : False ↔ True   := by rfl -- Error
example : P false true   := by rfl -- Error
example : Q false true   := by rfl -- Error
example : Q' false true  := by rfl -- Error
example : R false true   := by rfl -- Error

example : false = true   := by with_reducible rfl -- Error
example : HEq false true := by with_reducible rfl -- Error
example : False ↔ True   := by with_reducible rfl -- Error
example : P false true   := by with_reducible rfl -- Error
example : Q false true   := by with_reducible rfl -- Error
example : Q' false true  := by with_reducible rfl -- Error
example : R false true   := by with_reducible rfl -- Error

-- Inheterogeneous unequal

example : HEq true 1 := by rfl -- Error
example : HEq true 1 := by with_reducible rfl -- Error

-- Rfl lemma with side condition:
-- Error message should show left-over goal

inductive S : Bool → Bool → Prop where | refl : a = true → S a a
attribute [refl] S.refl
example : S true false  := by rfl -- Error
example : S true false  := by with_reducible rfl -- Error
example : S true true   := by rfl -- Error (left-over goal)
example : S true true   := by with_reducible rfl -- Error (left-over goal)
example : S false false := by rfl -- Error (left-over goal)
example : S false false := by with_reducible rfl -- Error (left-over goal)
