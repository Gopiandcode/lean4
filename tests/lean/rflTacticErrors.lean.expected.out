rflTacticErrors.lean:29:30-29:33: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:30:30-30:33: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:37:45-37:48: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:38:45-38:48: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:50:31-50:34: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:51:31-51:34: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:58:46-58:49: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:59:46-59:49: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:71:32-71:35: error: rfl failed, no @[refl] lemma registered for relation
  Q'
rflTacticErrors.lean:72:32-72:35: error: rfl failed, no @[refl] lemma registered for relation
  R
rflTacticErrors.lean:74:47-74:50: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ true'' = true
rflTacticErrors.lean:75:47-75:50: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq true'' true
⊢ HEq true'' true
rflTacticErrors.lean:76:47-76:50: error: tactic 'rfl' failed, The lhs
  True''
is not definitionally equal to rhs
  True
⊢ True'' ↔ True
rflTacticErrors.lean:77:47-77:50: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ P true'' true
rflTacticErrors.lean:78:47-78:50: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ Q true'' true
rflTacticErrors.lean:79:47-79:50: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ Q' true'' true
rflTacticErrors.lean:80:47-80:50: error: tactic 'rfl' failed, The lhs
  true''
is not definitionally equal to rhs
  true
⊢ R true'' true
rflTacticErrors.lean:83:31-83:34: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ false = true
rflTacticErrors.lean:84:31-84:34: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq false true
⊢ HEq false true
rflTacticErrors.lean:85:31-85:34: error: tactic 'rfl' failed, The lhs
  False
is not definitionally equal to rhs
  True
⊢ False ↔ True
rflTacticErrors.lean:86:31-86:34: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ P false true
rflTacticErrors.lean:87:31-87:34: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q false true
rflTacticErrors.lean:88:31-88:34: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q' false true
rflTacticErrors.lean:89:31-89:34: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ R false true
rflTacticErrors.lean:91:46-91:49: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ false = true
rflTacticErrors.lean:92:46-92:49: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq false true
⊢ HEq false true
rflTacticErrors.lean:93:46-93:49: error: tactic 'rfl' failed, The lhs
  False
is not definitionally equal to rhs
  True
⊢ False ↔ True
rflTacticErrors.lean:94:46-94:49: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ P false true
rflTacticErrors.lean:95:46-95:49: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q false true
rflTacticErrors.lean:96:46-96:49: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ Q' false true
rflTacticErrors.lean:97:46-97:49: error: tactic 'rfl' failed, The lhs
  false
is not definitionally equal to rhs
  true
⊢ R false true
rflTacticErrors.lean:101:27-101:30: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq true 1
⊢ HEq true 1
rflTacticErrors.lean:102:42-102:45: error: tactic 'apply' failed, failed to unify
  HEq ?a ?a
with
  HEq true 1
⊢ HEq true 1
rflTacticErrors.lean:109:30-109:33: error: tactic 'rfl' failed, The lhs
  true
is not definitionally equal to rhs
  false
⊢ S true false
rflTacticErrors.lean:110:45-110:48: error: tactic 'rfl' failed, The lhs
  true
is not definitionally equal to rhs
  false
⊢ S true false
rflTacticErrors.lean:111:30-111:33: error: unsolved goals
case a
⊢ true = true
rflTacticErrors.lean:112:45-112:48: error: unsolved goals
case a
⊢ true = true
rflTacticErrors.lean:113:30-113:33: error: unsolved goals
case a
⊢ false = true
rflTacticErrors.lean:114:45-114:48: error: unsolved goals
case a
⊢ false = true
